File: examples/websocket_client.py
"""
WebSocket Client Example - Connect to backtest streaming

This example demonstrates how to connect to the Stock Picker backtest
WebSocket API and receive real-time updates.

Usage:
    python examples/websocket_client.py --run-id <backtest_run_id>
    python examples/websocket_client.py --all  # Listen to all backtests
    python examples/websocket_client.py --stats  # Connection stats

Requirements:
    pip install websockets asyncio
"""

import asyncio
import json
import argparse
from datetime import datetime
from typing import Optional

try:
    import websockets
except ImportError:
    print("Error: websockets library not installed")
    print("Install with: pip install websockets")
    exit(1)


class BacktestWebSocketClient:
    """WebSocket client for backtest streaming"""

    def __init__(self, base_url: str = "ws://localhost:8000"):
        self.base_url = base_url
        self.websocket: Optional[websockets.WebSocketClientProtocol] = None

    async def connect_to_backtest(self, run_id: str, client_id: Optional[str] = None):
        """
        Connect to a specific backtest stream

        Args:
            run_id: Backtest run ID
            client_id: Optional client identifier
        """
        url = f"{self.base_url}/api/v1/ws/backtest/{run_id}"
        if client_id:
            url += f"?client_id={client_id}"

        print(f"Connecting to backtest stream: {run_id}")
        print(f"URL: {url}")
        print("-" * 60)

        async with websockets.connect(url) as websocket:
            self.websocket = websocket

            # Receive connection acknowledgment
            message = await websocket.recv()
            data = json.loads(message)
            print(f"✓ Connected: {data.get('message', 'N/A')}")
            print(f"  Client ID: {data.get('client_id', 'N/A')}")
            print("-" * 60)

            # Listen for events
            try:
                async for message in websocket:
                    data = json.loads(message)
                    self._handle_message(data)

            except websockets.exceptions.ConnectionClosed:
                print("\n✗ Connection closed")

    async def connect_to_all_backtests(self, client_id: Optional[str] = None):
        """
        Connect to all backtests stream

        Args:
            client_id: Optional client identifier
        """
        url = f"{self.base_url}/api/v1/ws/backtests"
        if client_id:
            url += f"?client_id={client_id}"

        print(f"Connecting to all backtests stream")
        print(f"URL: {url}")
        print("-" * 60)

        async with websockets.connect(url) as websocket:
            self.websocket = websocket

            # Receive connection acknowledgment
            message = await websocket.recv()
            data = json.loads(message)
            print(f"✓ Connected: {data.get('message', 'N/A')}")
            print("-" * 60)

            # Listen for events
            try:
                async for message in websocket:
                    data = json.loads(message)
                    self._handle_message(data)

            except websockets.exceptions.ConnectionClosed:
                print("\n✗ Connection closed")

    async def connect_to_stats(self, client_id: Optional[str] = None):
        """
        Connect to connection stats stream

        Args:
            client_id: Optional client identifier
        """
        url = f"{self.base_url}/api/v1/ws/stats"
        if client_id:
            url += f"?client_id={client_id}"

        print(f"Connecting to stats stream")
        print(f"URL: {url}")
        print("-" * 60)

        async with websockets.connect(url) as websocket:
            self.websocket = websocket

            # Request stats every 5 seconds
            try:
                while True:
                    # Send get_stats command
                    await websocket.send(json.dumps({
                        "command": "get_stats",
                        "timestamp": datetime.now().isoformat()
                    }))

                    # Receive response
                    message = await websocket.recv()
                    data = json.loads(message)
                    self._handle_message(data)

                    # Wait 5 seconds
                    await asyncio.sleep(5)

            except websockets.exceptions.ConnectionClosed:
                print("\n✗ Connection closed")

    def _handle_message(self, data: dict):
        """
        Handle incoming WebSocket message

        Args:
            data: Parsed JSON message
        """
        msg_type = data.get("type", "unknown")
        timestamp = data.get("timestamp", "N/A")
        run_id = data.get("run_id", "N/A")

        if msg_type == "progress":
            # Progress update
            progress_data = data.get("data", {})
            current = progress_data.get("current", 0)
            total = progress_data.get("total", 1)
            percentage = progress_data.get("percentage", 0)
            message = progress_data.get("message", "")

            print(f"📊 Progress: {percentage:.1f}% ({current}/{total})")
            if message:
                print(f"   {message}")

        elif msg_type == "trade":
            # Trade event
            trade_data = data.get("data", {})
            trade_type = trade_data.get("trade_type", "unknown")
            symbol = trade_data.get("symbol", "N/A")
            price = trade_data.get("price", 0)
            quantity = trade_data.get("quantity", 0)

            emoji = "🟢" if trade_type == "entry" else "🔴"
            print(f"{emoji} Trade {trade_type.upper()}: {symbol} @ {price} x {quantity}")

        elif msg_type == "snapshot":
            # Portfolio snapshot
            snapshot_data = data.get("data", {})
            total_value = snapshot_data.get("total_value", 0)
            cash = snapshot_data.get("cash", 0)
            positions = snapshot_data.get("positions", 0)

            print(f"💼 Portfolio: ${total_value:,.2f} (Cash: ${cash:,.2f}, Positions: {positions})")

        elif msg_type == "status":
            # Status change
            status_data = data.get("data", {})
            status = status_data.get("status", "unknown")
            message = status_data.get("message", "")

            emoji = "✓" if status == "completed" else "⚠" if status == "failed" else "▶"
            print(f"{emoji} Status: {status.upper()}")
            if message:
                print(f"   {message}")

        elif msg_type == "error":
            # Error event
            error_data = data.get("data", {})
            error = error_data.get("error", "Unknown error")
            details = error_data.get("details", {})

            print(f"❌ Error: {error}")
            if details:
                print(f"   Details: {details}")

        elif msg_type == "stats":
            # Connection stats
            stats_data = data.get("data", {})
            active_connections = stats_data.get("active_connections", 0)
            active_rooms = stats_data.get("active_rooms", 0)
            rooms = stats_data.get("rooms", [])

            print(f"📈 Connection Stats:")
            print(f"   Active Connections: {active_connections}")
            print(f"   Active Rooms: {active_rooms}")
            print(f"   Rooms: {', '.join(rooms) if rooms else 'None'}")

        elif msg_type in ["backtest_started", "backtest_completed", "backtest_failed"]:
            # Global backtest events
            status_data = data.get("data", {})
            message = status_data.get("message", "")

            emoji = "🚀" if msg_type == "backtest_started" else "✅" if msg_type == "backtest_completed" else "❌"
            print(f"{emoji} {msg_type.replace('_', ' ').title()}: {run_id}")
            if message:
                print(f"   {message}")

        else:
            # Unknown message type
            print(f"📨 Message ({msg_type}):")
            print(f"   {json.dumps(data, indent=2)}")

        print()  # Blank line for readability


async def main():
    """Main entry point"""
    parser = argparse.ArgumentParser(description="WebSocket client for backtest streaming")
    parser.add_argument("--run-id", help="Backtest run ID to stream")
    parser.add_argument("--all", action="store_true", help="Listen to all backtests")
    parser.add_argument("--stats", action="store_true", help="Monitor connection stats")
    parser.add_argument("--url", default="ws://localhost:8000", help="WebSocket base URL")
    parser.add_argument("--client-id", help="Client identifier")

    args = parser.parse_args()

    client = BacktestWebSocketClient(base_url=args.url)

    try:
        if args.run_id:
            await client.connect_to_backtest(args.run_id, args.client_id)
        elif args.all:
            await client.connect_to_all_backtests(args.client_id)
        elif args.stats:
            await client.connect_to_stats(args.client_id)
        else:
            print("Error: Must specify --run-id, --all, or --stats")
            parser.print_help()
            return

    except KeyboardInterrupt:
        print("\n\n✓ Disconnected by user")
    except Exception as e:
        print(f"\n❌ Error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/websocket_client.js
/**
 * WebSocket Client Example (JavaScript/Node.js)
 *
 * This example demonstrates how to connect to the Stock Picker backtest
 * WebSocket API using JavaScript/Node.js.
 *
 * Usage:
 *     node examples/websocket_client.js --run-id <backtest_run_id>
 *     node examples/websocket_client.js --all
 *     node examples/websocket_client.js --stats
 *
 * Requirements:
 *     npm install ws
 */

const WebSocket = require('ws');

class BacktestWebSocketClient {
    constructor(baseUrl = 'ws://localhost:8000') {
        this.baseUrl = baseUrl;
        this.ws = null;
    }

    /**
     * Connect to a specific backtest stream
     */
    connectToBacktest(runId, clientId = null) {
        let url = `${this.baseUrl}/api/v1/ws/backtest/${runId}`;
        if (clientId) {
            url += `?client_id=${clientId}`;
        }

        console.log(`Connecting to backtest stream: ${runId}`);
        console.log(`URL: ${url}`);
        console.log('-'.repeat(60));

        this.ws = new WebSocket(url);

        this.ws.on('open', () => {
            console.log('✓ WebSocket connection opened');
        });

        this.ws.on('message', (data) => {
            const message = JSON.parse(data);
            this._handleMessage(message);
        });

        this.ws.on('close', () => {
            console.log('\n✗ Connection closed');
        });

        this.ws.on('error', (error) => {
            console.error('❌ WebSocket error:', error.message);
        });
    }

    /**
     * Connect to all backtests stream
     */
    connectToAllBacktests(clientId = null) {
        let url = `${this.baseUrl}/api/v1/ws/backtests`;
        if (clientId) {
            url += `?client_id=${clientId}`;
        }

        console.log('Connecting to all backtests stream');
        console.log(`URL: ${url}`);
        console.log('-'.repeat(60));

        this.ws = new WebSocket(url);

        this.ws.on('open', () => {
            console.log('✓ WebSocket connection opened');
        });

        this.ws.on('message', (data) => {
            const message = JSON.parse(data);
            this._handleMessage(message);
        });

        this.ws.on('close', () => {
            console.log('\n✗ Connection closed');
        });

        this.ws.on('error', (error) => {
            console.error('❌ WebSocket error:', error.message);
        });
    }

    /**
     * Connect to connection stats stream
     */
    connectToStats(clientId = null) {
        let url = `${this.baseUrl}/api/v1/ws/stats`;
        if (clientId) {
            url += `?client_id=${clientId}`;
        }

        console.log('Connecting to stats stream');
        console.log(`URL: ${url}`);
        console.log('-'.repeat(60));

        this.ws = new WebSocket(url);

        this.ws.on('open', () => {
            console.log('✓ WebSocket connection opened');

            // Request stats every 5 seconds
            setInterval(() => {
                if (this.ws.readyState === WebSocket.OPEN) {
                    this.ws.send(JSON.stringify({
                        command: 'get_stats',
                        timestamp: new Date().toISOString()
                    }));
                }
            }, 5000);
        });

        this.ws.on('message', (data) => {
            const message = JSON.parse(data);
            this._handleMessage(message);
        });

        this.ws.on('close', () => {
            console.log('\n✗ Connection closed');
        });

        this.ws.on('error', (error) => {
            console.error('❌ WebSocket error:', error.message);
        });
    }

    /**
     * Send ping command
     */
    sendPing() {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({
                command: 'ping',
                timestamp: new Date().toISOString()
            }));
        }
    }

    /**
     * Unsubscribe from current stream
     */
    unsubscribe() {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({
                command: 'unsubscribe'
            }));
        }
    }

    /**
     * Handle incoming WebSocket message
     */
    _handleMessage(data) {
        const msgType = data.type || 'unknown';
        const timestamp = data.timestamp || 'N/A';
        const runId = data.run_id || 'N/A';

        switch (msgType) {
            case 'connected':
                console.log(`✓ Connected: ${data.message || 'N/A'}`);
                console.log(`  Client ID: ${data.client_id || 'N/A'}`);
                console.log('-'.repeat(60));
                break;

            case 'progress':
                const progressData = data.data || {};
                const percentage = progressData.percentage || 0;
                const current = progressData.current || 0;
                const total = progressData.total || 1;
                const message = progressData.message || '';

                console.log(`📊 Progress: ${percentage.toFixed(1)}% (${current}/${total})`);
                if (message) {
                    console.log(`   ${message}`);
                }
                break;

            case 'trade':
                const tradeData = data.data || {};
                const tradeType = tradeData.trade_type || 'unknown';
                const symbol = tradeData.symbol || 'N/A';
                const price = tradeData.price || 0;
                const quantity = tradeData.quantity || 0;

                const emoji = tradeType === 'entry' ? '🟢' : '🔴';
                console.log(`${emoji} Trade ${tradeType.toUpperCase()}: ${symbol} @ ${price} x ${quantity}`);
                break;

            case 'snapshot':
                const snapshotData = data.data || {};
                const totalValue = snapshotData.total_value || 0;
                const cash = snapshotData.cash || 0;
                const positions = snapshotData.positions || 0;

                console.log(`💼 Portfolio: $${totalValue.toLocaleString()} (Cash: $${cash.toLocaleString()}, Positions: ${positions})`);
                break;

            case 'status':
                const statusData = data.data || {};
                const status = statusData.status || 'unknown';
                const statusMessage = statusData.message || '';

                const statusEmoji = status === 'completed' ? '✓' : status === 'failed' ? '⚠' : '▶';
                console.log(`${statusEmoji} Status: ${status.toUpperCase()}`);
                if (statusMessage) {
                    console.log(`   ${statusMessage}`);
                }
                break;

            case 'error':
                const errorData = data.data || {};
                const error = errorData.error || 'Unknown error';
                const details = errorData.details || {};

                console.log(`❌ Error: ${error}`);
                if (Object.keys(details).length > 0) {
                    console.log(`   Details: ${JSON.stringify(details)}`);
                }
                break;

            case 'stats':
                const statsData = data.data || {};
                const activeConnections = statsData.active_connections || 0;
                const activeRooms = statsData.active_rooms || 0;
                const rooms = statsData.rooms || [];

                console.log('📈 Connection Stats:');
                console.log(`   Active Connections: ${activeConnections}`);
                console.log(`   Active Rooms: ${activeRooms}`);
                console.log(`   Rooms: ${rooms.length > 0 ? rooms.join(', ') : 'None'}`);
                break;

            case 'pong':
                console.log('🏓 Pong received');
                break;

            case 'unsubscribed':
                console.log('✓ Unsubscribed from stream');
                break;

            case 'backtest_started':
            case 'backtest_completed':
            case 'backtest_failed':
                const backtestEmoji = msgType === 'backtest_started' ? '🚀' : msgType === 'backtest_completed' ? '✅' : '❌';
                const backtestData = data.data || {};
                const backtestMessage = backtestData.message || '';

                console.log(`${backtestEmoji} ${msgType.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase())}: ${runId}`);
                if (backtestMessage) {
                    console.log(`   ${backtestMessage}`);
                }
                break;

            default:
                console.log(`📨 Message (${msgType}):`);
                console.log(JSON.stringify(data, null, 2));
        }

        console.log(); // Blank line for readability
    }

    /**
     * Close the WebSocket connection
     */
    close() {
        if (this.ws) {
            this.ws.close();
        }
    }
}

// Main entry point
function main() {
    const args = process.argv.slice(2);
    const client = new BacktestWebSocketClient();

    // Parse arguments
    let runId = null;
    let showAll = false;
    let showStats = false;
    let clientId = null;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--run-id' && i + 1 < args.length) {
            runId = args[i + 1];
            i++;
        } else if (args[i] === '--all') {
            showAll = true;
        } else if (args[i] === '--stats') {
            showStats = true;
        } else if (args[i] === '--client-id' && i + 1 < args.length) {
            clientId = args[i + 1];
            i++;
        }
    }

    // Connect based on arguments
    if (runId) {
        client.connectToBacktest(runId, clientId);
    } else if (showAll) {
        client.connectToAllBacktests(clientId);
    } else if (showStats) {
        client.connectToStats(clientId);
    } else {
        console.log('Usage:');
        console.log('  node examples/websocket_client.js --run-id <backtest_run_id>');
        console.log('  node examples/websocket_client.js --all');
        console.log('  node examples/websocket_client.js --stats');
        console.log('');
        console.log('Options:');
        console.log('  --client-id <id>    Optional client identifier');
        process.exit(1);
    }

    // Handle Ctrl+C
    process.on('SIGINT', () => {
        console.log('\n\n✓ Disconnected by user');
        client.close();
        process.exit(0);
    });
}

// Run if executed directly
if (require.main === module) {
    main();
}

module.exports = BacktestWebSocketClient;
